apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts

```diff
@@ -305,6 +331,209 @@ export class TypeScriptBuilder extends SubprocessRunnerBase<ITypeScriptBuilderCo
+    const createMultiEmitBuilderProgram: TTypescript.CreateProgram<
+      TTypescript.EmitAndSemanticDiagnosticsBuilderProgram
+    > = (
+      fileNames: readonly string[] | undefined,
+      compilerOptions: TTypescript.CompilerOptions | undefined,
+      host: TTypescript.CompilerHost | undefined,
+      oldProgram: TTypescript.EmitAndSemanticDiagnosticsBuilderProgram | undefined,
+      configFileParsingDiagnostics: readonly TTypescript.Diagnostic[] | undefined,
+      projectReferences: readonly TTypescript.ProjectReference[] | undefined
+    ): TTypescript.EmitAndSemanticDiagnosticsBuilderProgram => {
+      // Reset performance counters
+      ts.performance.disable();
+      ts.performance.enable();
+
+      this._typescriptTerminal.writeVerboseLine(`Reading program "${compilerOptions!.configFilePath}"`);
+
+      const newProgram: TTypescript.EmitAndSemanticDiagnosticsBuilderProgram =
+        ts.createEmitAndSemanticDiagnosticsBuilderProgram(
+          fileNames,
+          compilerOptions,
+          host,
+          oldProgram,
+          configFileParsingDiagnostics,
+          projectReferences
+        );
+
+      this._logReadPerformance(ts);
+
+      const { emit: originalEmit } = newProgram;
+
+      const emit: TTypescript.Program['emit'] = (
+        outerTargetSourceFile?: TTypescript.SourceFile,
+        outerWriteFile?: TTypescript.WriteFileCallback,
+        outerCancellationToken?: TTypescript.CancellationToken,
+        outerEmitOnlyDtsFiles?: boolean,
+        outerCustomTransformers?: TTypescript.CustomTransformers
+      ) => {
+        const innerProgram: TTypescript.Program = newProgram.getProgram();
+
+        const innerCompilerOptions: TTypescript.CompilerOptions = innerProgram.getCompilerOptions();
+
+        const { changedFiles } = this._configureProgramForMultiEmit(innerProgram, ts);
+
+        const result: TTypescript.EmitResult = originalEmit.call(
+          newProgram,
+          outerTargetSourceFile,
+          outerWriteFile,
+          outerCancellationToken,
+          outerEmitOnlyDtsFiles,
+          outerCustomTransformers
+        );
+
+        (result as IExtendedEmitResult).changedSourceFiles = changedFiles;
+
+        this._typescriptTerminal.writeVerboseLine(
+          `Emitting program "${innerCompilerOptions!.configFilePath}"`
+        );
+
+        this._logEmitPerformance(ts);
+
+        // Reset performance counters
+        ts.performance.disable();
+        ts.performance.enable();
+
+        return result;
+      };
+
+      newProgram.emit = emit;
+
+      return newProgram;
+    };
+
+    return createMultiEmitBuilderProgram;
+  }
+
+  private _configureProgramForMultiEmit(
+    innerProgram: TTypescript.Program,
+    ts: ExtendedTypeScript
+  ): { changedFiles: Set<IExtendedSourceFile> } {
+    interface IProgramWithMultiEmit extends TTypescript.Program {
+      __innerGetCompilerOptions?: TTypescript.Program['getCompilerOptions'];
```

Comment:
If these are undocumented properties on the program, add a comment permalinking to the TS repo where they are defined. If they're something that you're adding yourself (which is what it appears to be), add some comments describing what you're doing here